tests/fixtures/test_logger.h

```diff
@@ -81,6 +118,64 @@ class TestLogger : public vsag::Logger {
     std::mutex mutex_;
 };
 
-extern TestLogger logger;
+class LoggerStream : public std::basic_streambuf<char> {
+public:
+    explicit LoggerStream(TestLogger* logger,
+                          vsag::Logger::Level level,
+                          uint64_t buffer_size = 1024)
+        : logger_(logger), level_(level), buffer_(buffer_size + 1) {
+        auto base = &buffer_.front();
+        this->setp(base, base + buffer_size);
+    }
+
+    virtual ~LoggerStream() {
+        logger_ = nullptr;
+    }
+
+public:
+    virtual int
+    overflow(int ch) override {
+        std::lock_guard<std::mutex> lock(mutex_);
+        if (ch != EOF) {
+            *this->pptr() = (char)ch;
+            this->pbump(1);
+        }
+        this->flush();
+        return ch;
+    }
+
+    virtual int
+    sync() override {
+        std::lock_guard<std::mutex> lock(mutex_);
+        this->flush();
+        return 0;
+    }
+
+private:
+    void
+    flush() {
+        std::ptrdiff_t n = this->pptr() - this->pbase();
+        std::string msg(this->pbase(), n);
+        this->pbump(-n);
+        if (logger_) {
+            logger_->Log(msg, level_);
+        }
+    }
+
+private:
+    TestLogger* logger_ = nullptr;
+    vsag::Logger::Level level_;
+    std::mutex mutex_;
+    std::vector<char> buffer_;
+    uint64_t size_;
+};
+
+extern TestLogger test_logger;
+extern std::basic_ostream<char> trace;
+extern std::basic_ostream<char> debug;
+extern std::basic_ostream<char> info;
+extern std::basic_ostream<char> warn;
+extern std::basic_ostream<char> error;
+extern std::basic_ostream<char> critical;
```

Comment:
`logger::debug( ... )` provides better performance, but we prefer to use `logger::debug << ... ` in test codes to output variables simply.